src/materialized/src/http.rs

```diff
@@ -285,91 +218,160 @@ impl Server {
 }
 
 #[derive(Clone)]
-pub struct ThirdPartyServer {
-    metrics_registry: MetricsRegistry,
+enum ConnDetails {
+    Http,
+    Https { client_cert: Option<X509> },
 }
 
-impl ThirdPartyServer {
-    pub fn new(metrics_registry: MetricsRegistry) -> Self {
-        Self { metrics_registry }
-    }
+struct AuthedUser(String);
+
+pub struct AuthedClient(pub SessionClient);
+
+#[async_trait]
+impl<B> FromRequest<B> for AuthedClient
+where
+    B: Send,
+{
+    type Rejection = (StatusCode, String);
+
+    async fn from_request(req: &mut RequestParts<B>) -> Result<Self, Self::Rejection> {
+        let AuthedUser(user) = req.extensions().get::<AuthedUser>().unwrap();
+        let coord_client = req.extensions().get::<mz_coord::Client>().unwrap();
+
+        let coord_client = coord_client
+            .new_conn()
+            .map_err(|e| (StatusCode::INTERNAL_SERVER_ERROR, e.to_string()))?;
+        let session = Session::new(coord_client.conn_id(), user.clone());
+        let (coord_client, _) = match coord_client.startup(session).await {
+            Ok(coord_client) => coord_client,
+            Err(e) => {
+                return Err((StatusCode::INTERNAL_SERVER_ERROR, e.to_string()));
+            }
+        };
 
-    pub async fn serve(self, addr: SocketAddr) {
-        if let Err(err) = hyper::Server::bind(&addr)
-            .serve(hyper::service::make_service_fn(|_| {
-                let server = self.clone();
-                async { Ok::<_, hyper::Error>(server) }
-            }))
-            .await
-        {
-            error!("error serving metrics endpoint: {}", err);
-        }
+        Ok(AuthedClient(coord_client))
     }
 }
 
-impl hyper::service::Service<Request<Body>> for ThirdPartyServer {
-    type Response = Response<Body>;
-    type Error = hyper::http::Error;
-    type Future =
-        Pin<Box<dyn futures::Future<Output = Result<Self::Response, Self::Error>> + Send>>;
-
-    fn poll_ready(
-        &mut self,
-        _: &mut std::task::Context<'_>,
-    ) -> std::task::Poll<Result<(), Self::Error>> {
-        std::task::Poll::Ready(Ok(()))
+#[derive(Debug, Error)]
+enum AuthError {
+    #[error("HTTPS is required")]
+    HttpsRequired,
+    #[error("invalid username in client certificate")]
+    InvalidCertUserName,
+    #[error("{0}")]
+    Frontegg(#[from] FronteggError),
+    #[error("missing authorization header")]
+    MissingHttpAuthentication,
+    #[error("{0}")]
+    MismatchedUser(&'static str),
+}
+
+impl IntoResponse for AuthError {
+    fn into_response(self) -> Response {
+        (StatusCode::UNAUTHORIZED, self.to_string()).into_response()
```

Comment:
I had purposefully wrapped all frontegg auth errors in a generic string as an attempt to reduce information returned to attackers. I'm not convinced it was useful, but it may have been. One option requested by the cloud team was to log the full error (https://github.com/MaterializeInc/materialize/issues/10879). I'm ambivalent about what we do, but this could be a slight regression.

https://github.com/MaterializeInc/materialize/blob/bc71058aef4b6e599edd135c2cc79e76d3f94f51/src/materialized/src/http.rs#L209

Reply:
Right, this makes sense to me! I've adjusted the implementation here to return the generic "unauthorized" string to the user, but log a warning to the console.